onnx/defs/sequence/defs.cc

```diff
@@ -0,0 +1,665 @@
+// Copyright (c) ONNX Project Contributors.
+// Licensed under the MIT license.
+
+#include "onnx/defs/function.h"
+#include "onnx/defs/schema.h"
+
+#include <numeric>
+#include <algorithm>
+
+namespace ONNX_NAMESPACE {
+
+// target-shape = Union (target-shape, source_shape)
+// Example 1: same rank, different dimensions
+//    input1 shape: (2, 3, 4, 'x')
+//    input2 shape: (2, 'y', 5, 'x')
+//    output shape: (2, None, None, 'x')
+// Example 2: different rank
+//    input1 shape: (2, 3, 4, 'x')
+//    input2 shape: (2, 3, 4)
+//    output shape: None
+void UnionShapeInfo(
+    const TensorShapeProto& source_shape,
+    TypeProto_Tensor& target_type) {
+  if (target_type.has_shape()) {
+    TensorShapeProto* target_shape = target_type.mutable_shape();
+
+    auto source_rank = source_shape.dim_size();
+    auto target_rank = target_shape->dim_size();
+    if (source_rank != target_rank) {
+      target_type.clear_shape();
+      return;
+    }
+
+    for (int i = 0; i < source_rank; ++i) {
+      const auto source_dim = source_shape.dim(i);
+      const auto target_dim = target_shape->dim(i);
+      bool is_dims_conflict = [&](){
+        if (source_dim.has_dim_value()) {
+          if (target_dim.has_dim_value() &&
+              target_dim.dim_value() == source_dim.dim_value()) {
+            return false;
+          }
+          return true;
+        }
+
+        if (source_dim.has_dim_param()) {
+          if (target_dim.has_dim_param() &&
+              target_dim.dim_param() == source_dim.dim_param()) {
+            return false;
+          }
+          return true;
+        }
+
+        return (target_dim.has_dim_value() || target_dim.has_dim_param());
+      }();
+      if (is_dims_conflict &&
+          (target_dim.has_dim_value() || target_dim.has_dim_param())) {
+        auto dim = target_shape->mutable_dim(i);
+        dim->clear_dim_value();
+        dim->clear_dim_param();
+      }
+    }
+  }
+}
+
+static const char* SequenceEmpty_ver11_doc = R"DOC(
+Construct an empty tensor sequence, with given data type.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    SequenceEmpty,
+    11,
+    OpSchema()
+        .SetDoc(SequenceEmpty_ver11_doc)
+        .Attr(
+            "dtype",
+            "(Optional) The data type of the tensors in the output sequence. "
+            "The default type is 'float'.",
+            AttributeProto::INT,
```

Comment:
It should be a TypeProto but I don't know if the current framework allows it.

Reply:
Agree. We don't have any existing examples for doing this. We could use a string-based language (as in the specification of input/output types of operators) for specifying this.

Reply:
The current proto doesn't support it. It's good to add it once we are extending the functionalities further to support nested sequences.